LUCENE-9024 Optimize IntroSelector to use median of medians

pcsanwald opened the pull request:
# Description

Today, `IntroSelector.slowSelect` falls back to a heap sort; This PR moves that implementation to use a median of medians approach, suggested by a TODO in the code.
 
# Solution

Implemented a median of medians algorithm.

# Tests

I didn't add any new tests for this as I believe the current test cases, `TestIntroSelector`, should cover the basic functionality. However, one thing I'd like input on: Today, `select(int from, int to, int k)`, the `to` parameter is _exclusive_. Since it's also a specified parameter in the test, the test never checks the case where the full array should be used. `doTestSelect` has a call to `Arrays.sort(expected, from, to);`, but `to` will never be the last index in the array because the size of the array is allocated to `from + to + random().nextInt(5)`, so the actual size of the array will be a minimum of `from + to`.

I'd be happy to address this problem in this PR, or in a subsequent PR if that's preferred. 

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I am authorized to contribute this code to the ASF and have removed any code I do not have a license to distribute.
- [ ] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `ant precommit` and the appropriate test suite.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

jpountz commented on lucene/core/src/java/org/apache/lucene/util/IntroSelector.java:
```diff
@@ -33,26 +32,87 @@ public final void select(int from, int to, int k) {
+    return medianOfMediansSelect(from, to-1, k);
+  }
+
+  int medianOfMediansSelect(int from, int to, int k) {
```
It took me a while to figure out that `to` is inclusive here while it is exclusive elsewhere in this file (e.g. in `select` and `quickselecet`). Maybe rename it to `toInclusive` or something along those lines to make it clear? Or maybe call them `left` and `right` like you did in some other methods?

jpountz commented on lucene/core/src/java/org/apache/lucene/util/IntroSelector.java:
```diff
@@ -33,26 +32,87 @@ public final void select(int from, int to, int k) {
+  int medianOfMediansSelect(int from, int to, int k) {
+    int pivotIndex;
+    do {
+      if (from == to) {
```
I believe this condition can never happen when `medianOfMediansSelect` is called from `quickSelect` since it first checks whether `to-from==1` before delegating to `medianOfMediansSelect`? It is probably a good idea to be defensive, but maybe leave a comment about it?

jpountz commented on lucene/core/src/java/org/apache/lucene/util/IntroSelector.java:
```diff
@@ -33,26 +32,87 @@ public final void select(int from, int to, int k) {
+        from = pivotIndex+1;
+      }
+    } while (from != to);
+    setPivot(pivotIndex);
```
Setting the pivot here doesn't seem necessary?

jpountz commented on lucene/core/src/java/org/apache/lucene/util/IntroSelector.java:
```diff
@@ -33,26 +32,87 @@ public final void select(int from, int to, int k) {
-      @Override
-      protected void swap(int i, int j) {
-        IntroSelector.this.swap(i, j);
+  private int partition(int left, int right, int n, int pivotIndex) {
```
can we rename `n` to `k`, I think it's confusing that the same thing gets called different ways?

jpountz commented on lucene/core/src/java/org/apache/lucene/util/IntroSelector.java:
```diff
@@ -33,26 +32,87 @@ public final void select(int from, int to, int k) {
 
-      public void sort(int from, int to) {
-        heapSort(from, to);
+  private int partition5(int left, int right) {
```
Is this a bubble sort? If so it might make things clearer to say it in a comment, and maybe also be explicit that it doesn't hurt the runtime complexity since we only apply it to small arrays?

pcsanwald commented:
@jpountz thanks for the detailed review! I will be addressing your comments today. One comment:

> it should be `to+random.nextInt(5)`, is it what you are proposing?

That's part of it. but even with a right parameter of `to+0`, the code is never exercised over the whole array, as it's exclusive. So, I am not 100% sure I am right, but, it seems to me that there is no way to quickSelect on an entire array, as `to = arr.len` is still exclusive. I guess this has not been a problem in practice, but, I wanted to raise it.
